backend/python-functions/data_ingest.py

```diff
@@ -0,0 +1,145 @@
+            "popTestedPositive": 2
+          },
+          "2020-06-03": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          },
+          "2020-06-04": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          }
+        }
+      },
+      ...
+    }
+    """
+    facilities = {}    
+    with open(file_location) as csv_file:
+        reader = csv.DictReader(csv_file, delimiter=',')
+
+        for row in reader:
+            state = row["State"].strip()
+            canonical_facility_name = row["Canonical Facility Name"].strip()
+            facility_type = row['Facility Type'].strip()
+            key = f'{state}::{canonical_facility_name}'
+            
+            if(key in facilities):
+                # If the facility already exists, append the case counts for a given day.
+                facilities[key]['covidCases'][row['Date']] = build_covid_case_counts(row)
+            else:
+                # If the facility does not already exist, save its metadata along with
+                # the case counts for a given day.
+                facilities[key] = {
+                    'canonicalName': canonical_facility_name,
+                    'facilityType': facility_type,
+                    'state': state,
+                    'covidCases': {
+                        f'{row["Date"]}': build_covid_case_counts(row)
+                    }
+                }
+
+    return facilities
+
+def persist(facilities):
+    db = firestore.Client()
+
+    facilities_collection = db.collection(REFERENCE_FACILITIES_COLLECTION_ID)
+                
+    for key, facility in facilities.items():
+        facilitiesQueryResult = facilities_collection \
+            .where('state', '==', f'{facility["state"]}') \
+            .where('canonicalName', '==', f'{facility["canonicalName"]}') \
+            .stream()
+
+        facilityDocuments = list(facilitiesQueryResult)       
+
+        if len(facilityDocuments) == 0 or len(facilityDocuments) == 1:
+            # Accesses an existing reference facility document or creates a new one if
+            # one does not exist.
+            facility_id = facilityDocuments[0].id if len(facilityDocuments) == 1 else None
+            facility_ref = facilities_collection.document(facility_id)
+            
+            # Remove the Covid case data so that it can be stored separately in its own
+            # sub-collection.
+            covid_cases = facility.pop('covidCases')
+            
+            batch = db.batch()
+            
+            batch.set(facility_ref, facility)    
+
+            for date, cases in covid_cases.items():
+                covidCasesOnDateRef = facility_ref.collection('covidCases').document(date)
+                batch.set(covidCasesOnDateRef, cases)
+            
+            batch.commit()
+        else:
+            print(f'Multiple Documents were returned for ' \
```

Comment:
Is this indicative of an error we should be raising? I imagine we'd still want to process everything else so this is fine, but want to double-check.

Reply:
Yes, this represents a bad data case since at the moment we're forced to key off of the combination of the facility's state and name.  In other words, within a state, all facility names must be unique.  If this doesn't hold true for some reason, we wouldn't know which facility we should be updating so instead of guessing, we just log this message.

I've asked the data team to generate globally unique id's for each of the facilities so that we can avoid this and other more serious data issues.  [This appears to be in the works](https://docs.google.com/spreadsheets/d/1v50osSKG-vwmIhjIn6lFKDeY0ARsUT2R4olKsYsba2o/edit?disco=AAAAGnllJaU).

Reply:
I would recommend that you use a logger here instead of printing; that way you can give it ERROR status in the logs to make it easier to identify and respond to. (it's super simple, you can see an example in `helpers.py` if you are not familiar with this module in python)

Reply:
Great call! 🙌 

I'll update this `print` statement to use a logger.  Thanks for helping me by pointing out an example!